tests/phpunit/elementor/modules/image-loading-optimization/test-module.php

```diff
@@ -5,35 +5,64 @@
 		$document = self::factory()->create_post();
-		$document->update_main_meta( '_wp_page_template', $page_templates_module::TEMPLATE_CANVAS );
-
 		$content = '<img width="800" height="530" src="featured_image.jpg" /><img width="640" height="471" src="image_1.jpg" /><img width="800" height="800" src="image_2.jpg" /><img width="566" height="541" src="image_3.jpg" /><img width="691" height="1024" src="image_4.jpg" />';
 
-		// Update the post
-		$updated_post = array(
-			'ID'           => $document->get_main_id(),
-			'post_content' => $content,
-		);
-
-		// Update the post into the database
+		// Update the post content.
+		$updated_post = array( 'ID' => $document->get_main_id(), 'post_content' => $content );
 		wp_update_post( $updated_post );
+		$page_templates_module = Plugin::$instance->modules_manager->get_modules( 'page-templates' );
+		$document->update_main_meta( '_wp_page_template', $page_template );
 		
-		// Simulate a singular query.
 		query_posts( [ 'p' => $document->get_main_id() ] );
+		
+		// Need to do the_post as `template_include` depends on `wp_query` for the post.
+		the_post();
+		
+		// get template path from post. If it's not set switch to default single page template.
+		$template_path = locate_template('single.php');
+		$template_path = $page_templates_module->template_include( $template_path );
+
+		// Template usually contains a loop so need to rewind it.
+		rewind_posts();
+
 		ob_start();
-		$page_templates_module->print_content();
+		load_template($template_path);
+		/* php endscript works differently for phpunit.
+		 * This is to ensure buffer started in footer is closed.
+		 */ 
+		$this->close_and_print_open_buffer();
 		$output = ob_get_clean();
+
 		$expected = '<p><img fetchpriority="high" decoding="async" width="800" height="530" src="featured_image.jpg" /><img decoding="async" width="640" height="471" src="image_1.jpg" /><img decoding="async" width="800" height="800" src="image_2.jpg" /><img loading="lazy" decoding="async" width="566" height="541" src="image_3.jpg" /><img loading="lazy" decoding="async" width="691" height="1024" src="image_4.jpg" /></p>';
-		$this->assertStringContainsString( $expected, $output, "Loading Optimization not applied");
+		$this->assertStringContainsString( $expected, $output, "Loading optimization not applied to the content");
+	}
+
+	public function get_page_template() {
+		$page_templates_module = Plugin::$instance->modules_manager->get_modules( 'page-templates' );
+		return [
+			[ $page_templates_module::TEMPLATE_CANVAS ],
+			[ $page_templates_module::TEMPLATE_HEADER_FOOTER ],
+			[ $page_templates_module::TEMPLATE_THEME ]
+		];
+	}
+	
+	/**
+	 * This is to ensure buffer started in footer is closed.
+	 */
+	private function close_and_print_open_buffer() {
+		$buffer_status = ob_get_status();
+		if( ! empty( $buffer_status ) && str_contains( $buffer_status['name'], 'ImageLoadingOptimization\Module::handle_buffer_content') ) {
```

Comment:
This works as long as we don't change the callback name for that buffer. It makes me wonder if we should somehow try to find a way to close the footer buffer rather than relying on PHP to do so. Not something we have to solve for now, regardless. Let's get this in so we can keep moving forward!

Reply:
Agreed. I tried using `shutdown` but it didn't work the first time. I'll check again once.